src/python-common/ceph/deployment/utils.py

```diff
@@ -30,3 +32,67 @@ def is_ipv6(address):
         return ipaddress.ip_address(address).version == 6
     except ValueError:
         return False
+
+
+def valid_addr(addr: str) -> Tuple[bool, str]:
+    """check that an address string is valid
+    Valid in this case means that a name is resolvable, or the
+    IP address string is a correctly formed IPv4 or IPv6 address,
+    with or without an port
+
+    Args:
+        addr (str): address
+
+    Returns:
+        Tuple[bool, str]: Validity of the address, either
+                          True, address type (IPv4[:Port], IPv6[:Port], Name[:Port])
+                          False, <error description>
+    """
+
+    def _dns_lookup(addr: str, port: Optional[str]) -> Tuple[bool, str]:
+        try:
+            socket.getaddrinfo(addr, None)
+        except socket.gaierror:
+            # not resolvable
+            return False, 'DNS lookup failed'
+        return True, 'Name:Port' if port else 'Name'
+
+    def _ip_lookup(addr: str, port: Optional[str]) -> Tuple[bool, str]:
+        try:
+            ip_addr = ipaddress.ip_address(addr)
+        except ValueError:
+            return False, 'Invalid IP v4 or v6 address format'
+        return True, f'IPv{ip_addr.version}:Port' if port else f'IPv{ip_addr.version}'
+
+    brackets = addr.count('[') + addr.count(']')
+    dots = addr.count('.')
+    port = None
+
+    if addr.count(':') == 1:
+        addr, port = addr.split(':')  # drop the port
+    elif addr.startswith('[') and ']:' in addr:
+        port_found = addr.rfind(']:')
+        port = addr[port_found+2:]
+        addr = addr[1:port_found]  # drop the enclosing []:<port>
+    elif addr.startswith('[') and addr.endswith(']'):
+        addr = addr[1:-1]
+        if dots:
+            return False, "IPv4 address wrapped in brackets is invalid"
```

Comment:
We have this already:

https://github.com/ceph/ceph/blob/602cbea5ed38f2120e5b31d35c24774331601448/src/pybind/mgr/cephadm/services/monitoring.py#L138-L140

I don't think we need to parse urls manually.

Reply:
I've updated to use urlparse